packages/@lwc/engine-core/src/framework/rendering.ts

```diff
@@ -882,18 +905,37 @@ function updateDynamicChildren(
         } else if (isSameVnode(oldStartVnode, newEndVnode)) {
             // Vnode moved right
             patch(oldStartVnode, newEndVnode, parent, renderer);
-            insertNode(
-                oldStartVnode.elm!,
-                parent,
-                renderer.nextSibling(oldEndVnode.elm!),
-                renderer
-            );
+
+            // In the case of fragments, the `elm` property of a vfragment points to the leading
+            // anchor. To determine the next sibling of the whole fragment, we need to use the
+            // trailing anchor as the argument to nextSibling():
+            // [..., [leading, ...content, trailing], nextSibling, ...]
+            let anchor: Node | null;
+            if (isVFragment(oldEndVnode)) {
+                const trailing = oldEndVnode.children[oldEndVnode.children.length - 1]!.elm!;
```

Comment:
I think storing the trailing anchor node directly on the `vnode` could be a bit more elegant than implicitly getting it from the last child... but it's not a big deal if it's a pain to refactor.